File: src/main/java/com/springdebugger/extractor/TestResultsLocator.java
package com.springdebugger.extractor;

import java.io.File;
import java.util.ArrayList;
import java.util.List;

/**
 * Finds JUnit result XML files under a project tree: Gradle's {@code build/test-results/**.xml} and
 * Maven Surefire/Failsafe's {@code target/surefire-reports/*.xml} / {@code target/failsafe-reports}.
 *
 * <p>Walks with a bounded depth and prunes directories that never hold build output (source, VCS,
 * caches) so it stays cheap on large multi-module monorepos. Pure file-system logic, unit-tested
 * against a temporary tree.
 */
public final class TestResultsLocator {

    private static final int MAX_DEPTH = 8;

    private TestResultsLocator() {}

    public static List<File> locate(File projectBase) {
        List<File> results = new ArrayList<>();
        if (projectBase == null || !projectBase.isDirectory()) return results;
        walk(projectBase, 0, results);
        return results;
    }

    private static void walk(File dir, int depth, List<File> out) {
        if (depth > MAX_DEPTH) return;
        File[] children = dir.listFiles();
        if (children == null) return;
        for (File child : children) {
            if (child.isDirectory()) {
                String name = child.getName();
                if (isPruned(name)) continue;
                if (isResultDir(child)) {
                    collectXml(child, out);
                    // Result dirs do not nest result dirs; no need to descend further into them.
                    continue;
                }
                walk(child, depth + 1, out);
            }
        }
    }

    /** A directory whose contents are JUnit XML: Gradle test-results leaves and Maven report dirs. */
    private static boolean isResultDir(File dir) {
        String path = dir.getPath().replace(File.separatorChar, '/');
        return path.contains("/build/test-results/")
                || path.endsWith("/surefire-reports")
                || path.endsWith("/failsafe-reports");
    }

    private static boolean isPruned(String name) {
        switch (name) {
            case ".git":
            case ".gradle":
            case ".idea":
            case "node_modules":
            case "src":
            case ".m2":
                return true;
            default:
                return name.startsWith(".") && !name.equals(".");
        }
    }

    private static void collectXml(File dir, List<File> out) {
        File[] files = dir.listFiles();
        if (files == null) return;
        for (File f : files) {
            if (f.isFile() && f.getName().endsWith(".xml")) {
                out.add(f);
            } else if (f.isDirectory()) {
                // Gradle nests per-suite XML one level down (build/test-results/test/*.xml).
                collectXml(f, out);
            }
        }
    }
}

File: vscode-extension/src/engine/test-results-locator.ts
// Port of com.springdebugger.extractor.TestResultsLocator. Finds JUnit result XML under a project
// tree (Gradle build/test-results, Maven surefire/failsafe reports) with a bounded, pruned walk so
// it stays cheap on large monorepos. Returns absolute file paths.
import * as fs from 'fs';
import * as path from 'path';

const MAX_DEPTH = 8;

export function locate(projectBase: string): string[] {
  const results: string[] = [];
  if (!isDirectory(projectBase)) return results;
  walk(projectBase, 0, results);
  return results;
}

function walk(dir: string, depth: number, out: string[]): void {
  if (depth > MAX_DEPTH) return;
  for (const child of listDirs(dir)) {
    const name = path.basename(child);
    if (isPruned(name)) continue;
    if (isResultDir(child)) {
      collectXml(child, out);
      continue; // result dirs do not nest result dirs
    }
    walk(child, depth + 1, out);
  }
}

function isResultDir(dir: string): boolean {
  const p = dir.split(path.sep).join('/');
  return (
    p.includes('/build/test-results/') ||
    p.endsWith('/surefire-reports') ||
    p.endsWith('/failsafe-reports')
  );
}

function isPruned(name: string): boolean {
  switch (name) {
    case '.git':
    case '.gradle':
    case '.idea':
    case 'node_modules':
    case 'src':
    case '.m2':
      return true;
    default:
      return name.startsWith('.') && name !== '.';
  }
}

function collectXml(dir: string, out: string[]): void {
  let entries: fs.Dirent[];
  try {
    entries = fs.readdirSync(dir, { withFileTypes: true });
  } catch {
    return;
  }
  for (const e of entries) {
    const full = path.join(dir, e.name);
    if (e.isFile() && e.name.endsWith('.xml')) {
      out.push(full);
    } else if (e.isDirectory()) {
      collectXml(full, out); // Gradle nests per-suite XML one level down
    }
  }
}

function listDirs(dir: string): string[] {
  try {
    return fs
      .readdirSync(dir, { withFileTypes: true })
      .filter((e) => e.isDirectory())
      .map((e) => path.join(dir, e.name));
  } catch {
    return [];
  }
}

function isDirectory(p: string): boolean {
  try {
    return fs.statSync(p).isDirectory();
  } catch {
    return false;
  }
}
